### DISCUSSION TITLE
How to count consumed and inferred tokens?

### DISCUSSION THREAD

**Discussion by lostmsu** [SciSharp/LLamaSharp#922] [Q&A] [answered]
Subj

Currently I'm using `InteractiveExecutor` and `ChatSession` + `ChatAsync` to handle a single request (it is an API endpoint mimicking OpenAI)

Really though it might make more sense to do everything explicitly instead of using higher-level ops. I need to get initial tokens, count them, then generate new ones one by one, and stream back.

**Comment by martindevans** [comment, marked as answer]
If you'd prefer to work at the level of inidividual token, you're probably better off using the `BatchedExecutor`. That's a lower level executor (tokens in, logits out) so you have more control over the details (but it's harder to use than the higher level executors which are strings in, strings out).

**Reply by LSXAxeller** [reply]
is there a way to get the consumed tokens count and the remaining context length in the InteractiveExecutor?

**Reply by martindevans** [reply]
Not as far as I'm aware :(
